source/server/hot_restarting_parent.cc

```diff
@@ -0,0 +1,110 @@
+  socket_event_ = dispatcher.createFileEvent(
+      my_domain_socket(),
+      [this](uint32_t events) -> void {
+        ASSERT(events == Event::FileReadyType::Read);
+        onSocketEvent();
+      },
+      Event::FileTriggerType::Edge, Event::FileReadyType::Read);
+  server_ = &server;
+}
+
+void HotRestartingParent::onSocketEvent() {
+  std::unique_ptr<HotRestartMessage> wrapped_request;
+  while ((wrapped_request = receiveHotRestartMessage(Blocking::No))) {
+    if (wrapped_request->requestreply_case() == HotRestartMessage::kReply) {
+      ENVOY_LOG(error, "child sent us a reply type HotRestartMessage; ignoring.");
+      HotRestartMessage wrapped_reply;
+      wrapped_reply.set_didnt_recognize_your_last_message(true);
+      sendHotRestartMessage(child_address_, wrapped_reply);
+      continue;
+    }
+    switch (wrapped_request->request().request_case()) {
+    case HotRestartMessage::Request::kShutdownAdmin: {
+      server_->shutdownAdmin();
+      HotRestartMessage wrapped_reply;
+      wrapped_reply.mutable_reply()->mutable_shutdown_admin()->set_original_start_time_unix_seconds(
+          server_->startTimeFirstEpoch());
+      sendHotRestartMessage(child_address_, wrapped_reply);
+      break;
+    }
+
+    case HotRestartMessage::Request::kPassListenSocket: {
+      HotRestartMessage wrapped_reply;
+      wrapped_reply.mutable_reply()->mutable_pass_listen_socket()->set_fd(-1);
+      Network::Address::InstanceConstSharedPtr addr =
+          Network::Utility::resolveUrl(wrapped_request->request().pass_listen_socket().address());
+      for (const auto& listener : server_->listenerManager().listeners()) {
+        if (*listener.get().socket().localAddress() == *addr) {
+          wrapped_reply.mutable_reply()->mutable_pass_listen_socket()->set_fd(
+              listener.get().socket().ioHandle().fd());
+          break;
+        }
+      }
+      sendHotRestartMessage(child_address_, wrapped_reply);
+      break;
+    }
+
+    case HotRestartMessage::Request::kStats: {
+      HotRestartMessage wrapped_reply;
+      exportStatsToChild(wrapped_reply.mutable_reply()->mutable_stats());
+      sendHotRestartMessage(child_address_, wrapped_reply);
+      break;
+    }
+
+    case HotRestartMessage::Request::kDrainListeners: {
+      server_->drainListeners();
+      break;
+    }
+
+    case HotRestartMessage::Request::kTerminate: {
+      ENVOY_LOG(info, "shutting down due to child request");
+      kill(getpid(), SIGTERM);
+      break;
+    }
+
+    default: {
+      HotRestartMessage wrapped_reply;
+      wrapped_reply.set_didnt_recognize_your_last_message(true);
+      sendHotRestartMessage(child_address_, wrapped_reply);
+      break;
+    }
+    }
+  }
+}
+
+void HotRestartingParent::exportStatsToChild(HotRestartMessage::Reply::Stats* stats) {
+  for (const auto& gauge : server_->stats().gauges()) {
+    (*stats->mutable_gauges())[gauge->name()] = gauge->value();
```

Comment:
I think we're going to have a completely elaborated set of names in the keys of this map and counter_deltas, which is going to be quite significant when there are a large number of clusters.

At some point I thought the design of this was that we would stream out the deltas, so maybe we didn't need them all replicated in memory at once.

Can you leave a TODO to deal with this (or alternatively explain how this is not a memory high-water risk)?

Reply:
I meant to ask/comment on this before: @fredlas we should only send counters that have a latched value > 0 during the last interval. This will vastly reduce the amount of data sent between processes.

Reply:
Done.

Reply:
doesn't look done? Or is this coming?

Reply:
You may be seeing an old version; do you see `if (latched_value > 0)`?

Reply:
You are right; I was seeing the old version because github reviews :)

However even in the new version there's no TODO about somehow bounding the size of that map. Assume that every stat in a 10k cluster system is used, and a hot-restart is initiated. On the first iteration, a map will be created in memory containing all the fully elaborated stat names -- even though the underlying stats now contain StatName objects you will be elaboarting them to strings here and holding them all in memory at the same time. That seems potentially problematic.

Not asking you to fix it; just leave a TODO.

I think it could potentially be fixed by streaming out a bounded number of latched_value>0 counters per interval, or by just dividing into multiple messages of say (1k stats per message), each of which has a send/receive handshake. WDYT?

Reply:
Ah, sorry, I got derailed by the response and forgot about the first post. Yeah, it does seem like a problem, and sending chunk by chunk should fix it. Added the TODO.

Reply:
Yeah FWIW pagination is how I had originally thought about this, I think it's fine to leave as a TODO for now.

Reply:
@fredlas also, at the end of hot restart, this map is destroyed right?

Reply:
Correct. In fact, this particular map only exists for the duration of the processing of these periodic stats update messages.